### ISSUE TITLE
s/join-mode/pipe-mode/g

### ISSUE THREAD

**Issue by fopina** [semgrep/semgrep#6745]
Hi, not sure if this has been brought before (apologies if so) but I'd like to propose renaming join mode to pipe mode.

I assume the name comes either from SQL JOIN queries or from simply joining files, but the tool is called *grep which relates to shell.

Taking matches from a rule as output and using them as input for another rule would be piping (in a grep environment, meaning shell).

Just some 2 cents of fun!

**Comment by aryx** [comment]
True. Actually this join-mode is an experimental feature and we're planning to replace it with what we currently call multi-steps rules, that is the ability to have rules with multiple steps where each step leverage metavariables and matches from a previous step.

**Comment by hakangs** [comment]
@aryx is "multi-steps" currently integrated in the repo or is it smt that is being planned right now? also "multi-steps" will be open-source right (so different than DeepSemgrep)?

**Comment by aryx** [comment]
no it's not implemented. multi-steps would probably be open source, but it would be multiple steps inside the same file.
For multiple steps across files it would probably be only in DeepSemgrep.

**Comment by fopina** [comment]
But join mode is with different files right @aryx ?
So it will be “replaced” with a non-similar feature?

**Comment by aryx** [comment]
True, it would not be equivalent.
Join-mode is multiple things at the same time: it allows to combine/reuse multiple rules stored in different yaml files, it allows to
use one rule metavariables in another rule, it allows to analyze multiple files at the same time with multiple rules, and more probably. Pipe-mode would just allow to "use one rule metavariables in another rule". Jsonnet (another planned feature) would allow to combine/reuse multiple rules in different yaml files, which when combined with pipe-mode would cover some of the use cases of join-mode (but not all).

**Comment by fopina** [comment]
Thank you for the detailed info!
one common/easy use case of mine with join-mode is class inheritance in Java: flag all classes that extend A, even when C > B > A.
Is it correct to assume from your description that it won’t be possible with jsonnet?

**Comment by aryx** [comment]
Correct. This would require DeepSemgrep (now called Semgrep Pro).
